### ISSUE TITLE
expect(promise).to.eventually.deep.equal is always true when using with chai-stats

### ISSUE THREAD

**Issue by Getz85** [domenic/chai-as-promised#215]
Node: 6.11.1
Chai: 4.1.1
Chai-as-promised: 7.1.1
Mocha: 3.5.0
Q: 1.5.0

My tests based on 'eventually' are all green while they should fail.

This example above show that the tests with 'eventually' are not rejected, whereas they are rejected with a standard deep.equal:

```
var chai = require('chai'),
  expect = chai.expect,
  Q = require('Q'),
  chaiAsPromised = require('chai-as-promised');
  chai.use(chaiAsPromised);

  describe.only('testChaiAsPromised', () => {
    it('should fail with chai as promised with Q promise',
      function() {
        let deferred = Q.defer();
        deferred.resolve({
          category1: 90,
          category2: 120,
          total: 210
        });
        return expect(deferred.promise).to.eventually.deep.equal({
          category1: 100,
          category2: 150,
          total: 250
        });
      });

      it('should fail with chai as promised with Q promise and done callback',
        function(done) {
          let deferred = Q.defer();
          deferred.resolve({
            category1: 90,
            category2: 120,
            total: 210
          });
          expect(deferred.promise).to.eventually.deep.equal({
            category1: 100,
            category2: 150,
            total: 250
          }).and.notify(done);
        });

        it('should fail with Q promise and done callback',
          function(done) {
            let deferred = Q.defer();
            deferred.resolve({
              category1: 90,
              category2: 120,
              total: 210
            });
            deferred.promise.then(result => {
              expect(result).to.deep.equal({
                category1: 100,
                category2: 150,
                total: 250
              });
              done();
            }).catch(err => done(err));
          });
  });
```
The results:

```
testChaiAsPromised
      √ should fail with chai as promised with Q promise
      √ should fail with chai as promised with Q promise and done callback
      1) should fail with Q promise and done callback


  2 passing (62ms)
  1 failing

  1) services.alerts testChaiAsPromised should fail with Q promise and done callback:

      AssertionError: expected { Object (category1, category2, ...) } to deeply equal { Object (category1, category2, ...) }

      + expected - actual

       {
      -  "category1": 90
      -  "category2": 120
      -  "total": 210
      +  "category1": 100
      +  "category2": 150
      +  "total": 250
       }

      at deferred.promise.then.result (test\spec\integration\alerts.services.tests.js:86:38)
      at _fulfilled (node_modules\Q\q.js:854:54)
      at self.promiseDispatch.done (node_modules\Q\q.js:883:30)
      at Promise.promise.promiseDispatch (node_modules\Q\q.js:816:13)
      at node_modules\Q\q.js:570:49
      at runSingle (node_modules\Q\q.js:137:13)
      at flush (node_modules\Q\q.js:125:13)
      at _combinedTickCallback (internal/process/next_tick.js:73:7)
      at process._tickCallback (internal/process/next_tick.js:104:9)



npm ERR! Test failed.  See above for more details.
```

**Comment by domenic** [comment]
Does this problem occur with built-in promises?

Does this problem occur when not using Mocha (e.g. doing `expect(...)......then(console.log, console.error)`)?

**Comment by Getz85** [comment]
Well, after investigating a bit, it seems that the problems occurs when I use chai-stats and chai-as-promised in many tests, in separate files.

There is no problem if I launch a single test with just a require('chai-as-promised').

test1.js:
```
"use strict";
var chai = require('chai'),
  chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
let expect = chai.expect;
let Q = require('Q');
describe('testChaiAsPromised', () => {
  it('should fail with chai as promised with Q promise',
    function() {
      let deferred = Q.defer();
      deferred.resolve({
        category1: 90,
        category2: 120,
        total: 210
      });
      return expect(deferred.promise).to.eventually.deep.equal({
        category1: 100,
        category2: 150,
        total: 250
      });
    });

    it('should fail with chai as promised with Q promise and done callback',
      function(done) {
        let deferred = Q.defer();
        deferred.resolve({
          category1: 90,
          category2: 120,
          total: 210
        });
        expect(deferred.promise).to.eventually.deep.equal({
          category1: 100,
          category2: 150,
          total: 250
        }).and.notify(done);
      });

      it('should fail with Q promise and done callback',
        function(done) {
          let deferred = Q.defer();
          deferred.resolve({
            category1: 90,
            category2: 120,
            total: 210
          });
          deferred.promise.then(result => {
            expect(result).to.deep.equal({
              category1: 100,
              category2: 150,
              total: 250
            });
            done();
          }).catch(err => done(err));
        });
});
```
test2.js
```
"use strict";
var chai = require('chai'),
  chaiStats = require('chai-stats'),
  chaiAsPromised = require('chai-as-promised');
chai.use(chaiAsPromised);
chai.use(chaiStats);
var expect = chai.expect;

describe('testChaiAsPromised2', () => {
  it('should fail with chai as promised with native promise',
    function() {
      let promise = new Promise((resolve) => {
        resolve({category1: 90,
        category2: 120,
        total: 210
      });
      });
      return expect(promise).to.eventually.deep.equal({
        category1: 100,
        category2: 150,
        total: 250
      });
    });

    it('should fail with chai as promised with native promise and done callback',
      function(done) {
        let promise = new Promise((resolve) => {
          resolve({category1: 90,
          category2: 120,
          total: 210
        });
        });
        expect(promise).to.eventually.deep.equal({
          category1: 100,
          category2: 150,
          total: 250
        }).and.notify(done);
      });

      it('should fail with native promise and done callback',
        function(done) {
          let promise = new Promise((resolve) => {
            resolve({category1: 90,
            category2: 120,
            total: 210
          });
          });
          promise.then(result => {
            expect(result).to.deep.equal({
              category1: 100,
              category2: 150,
              total: 250
            });
            done();
          }).catch(err => done(err));
        });
});

```

And my file used to launch both tests files is just:

```
require('./test1.js');
require('./test2.js');
```

If you comment the line `chai.use(chaiStats);` in tests2.js, the tests fail as expected.
Maybe it's a chai or a chai-stats issue, not chai-as-promised one; but the behaviour of chai-as-promised seems to be 'parasite' by chaiStats.

Maybe should I `require('chai')` and make those `chai.use` in only one file, at launch, not for each file? Because adding chai-stats in test2.js make test1.js failed too...

**Comment by Getz85** [comment]
You can reproduce easily the issue with the tests in the repository, 33 tests failed.

Add ` "chai-stats": "^0.3.0"` in devDependencies.
And in setup.js:

```
"use strict";
const chai = require("chai");
const chaiAsPromised = require("../..");
const chaiStats = require('chai-stats');

chai.should();
chai.use(chaiAsPromised);
chai.use(chaiStats);

process.on("unhandledRejection", () => {
    // Do nothing; we test these all the time.
});
process.on("rejectionHandled", () => {
    // Do nothing; we test these all the time.
});
```

**Comment by domenic** [comment]
Thanks for your investigation. However, this isn't something I can really dig into on my own, as I don't use chai-stats. Furthermore, this reproduction cases seem to involve massive amounts of tests for me to read through and understand (100+ lines).

Let me know if you find something simple to fix; I'll be happy to accept a pull request.

**Comment by Getz85** [comment]
No pull-request, but a workaround if someone has the same issue.

Require only once chai, chai-stats and chai-as-promised for all the tests and use chai-stats BEFORE chai-as-promised.

```
var chai = require('chai');
chai.use(require('chai-stats'));
chai.use(require('chai-as-promised'));

global.expect = chai.expect;

require('./tmp.js');
require('./tmp2.js');
```

And in tmp.js and tmp2.js, just use the global.expect instead of `var expect = require('chai').expect;` :

```
let expect = global.expect;
```
